### ISSUE TITLE
delete-before-replace triggers delete-before-replace in all dependencies

### ISSUE THREAD

**Issue by hausdorff** [pulumi/pulumi#2167]
Following up from [this community Slack thread](https://pulumi-community.slack.com/archives/C84L4E3N1/p1541484105466000), there is a bug where:

* The user changes data in a manually-named Kubernetes `ConfigMap`, which triggers a delete-before-replace.
* The user references that `ConfigMap` in a `Deployment`.
* Because step generator [believes](https://github.com/pulumi/pulumi/blob/36c88aab3744ad2f18f6c264e58d34398e6794d4/pkg/resource/deploy/step_generator.go#L330) that any resource that is marked delete-before-replace must also have all its dependencies transitively marked delete-before-replace (specifically [this line](https://github.com/pulumi/pulumi/blob/36c88aab3744ad2f18f6c264e58d34398e6794d4/pkg/resource/deploy/step_generator.go#L353)), we delete the entire `Deployment`.

This is very bad for Kubernetes. It means that changing data in a `ConfigMap` can trigger a complete delete-first tear-down and rebuild of your application, including a stateful application.

The comment explaining this asserts that this is true, but I'm not really sure why it would be -- why, in particular, would we need to mark resources that actually don't need to be deleted at all, as delete-before-create?

cc @pgavlin @swgillespie @lukehoban

**Comment by lukehoban** [comment]
Here is the repro:

```typescript
import * as k8s from '@pulumi/kubernetes';
const someDataConfigMap = new k8s.core.v1.ConfigMap('some-config', {
  metadata: {
    name: 'some-config'
  },
  data: {
    // WHENEVER the value of foo is changed (i.e. change it to `bar2`) pulumi will first delete the deployment and then recreate it
    foo: 'bar1'
  }
});

const APP_NAME = 'foo';
const labels = {
  app: APP_NAME
};

const deployment = new k8s.apps.v1.Deployment(APP_NAME, {
  metadata: {
    name: APP_NAME
  },
  spec: {
    selector: {
      matchLabels: labels
    },
    template: {
      metadata: {
        labels
      },
      spec: {
        containers: [
          {
            name: 'main',
            image: 'nginx'
            // command: 'f'
          }
        ],
        volumes: [
          {
            name: 'some-data',
            configMap: {
              name: someDataConfigMap.metadata.apply(metadata => metadata.name)
            }
          }
        ]
      }
    }
  }
});
```

Changing `foo: "bar1"` to `foo: "bar2"` leads to the following:

```
$ pulumi update
Previewing update (dasdsadas):

     Type                           Name                 Plan        Info
     pulumi:pulumi:Stack            k8s-nginx-dasdsadas
 +-  ├─ kubernetes:apps:Deployment  foo                  replace
 +-  └─ kubernetes:core:ConfigMap   some-config          replace     [diff: ~data]

Resources:
    2 changes
    +-2 to replace
    1 unchanged
```

**Comment by lukehoban** [comment]
@hausdorff Why does changing the `data` property of a ConfigMap cause a replacement?  That property is allowed to change in the Kubernetes API without replacing the resource, right?

**Comment by hausdorff** [comment]
@lukehoban step generator [believes](https://github.com/pulumi/pulumi/blob/36c88aab3744ad2f18f6c264e58d34398e6794d4/pkg/resource/deploy/step_generator.go#L330) that any resource that is marked delete-before-replace must also have all its dependencies transitively marked delete-before-replace (specifically [this line](https://github.com/pulumi/pulumi/blob/36c88aab3744ad2f18f6c264e58d34398e6794d4/pkg/resource/deploy/step_generator.go#L353)).

I expand a bit on this in the initial issue.

**Comment by pgavlin** [comment]
The behavior that this issue describes is due to three intersecting features:
1. Pulumi's delete semantics
2. The Kubernetes provider's delete-before-replace semantics for resources with fixed names
3. The Kubernetes provider's replacement semantics for `ConfigMap`s

In reverse order:
- The Kubernetes provider treats `ConfigMap`s as immutable in order to trigger rollouts when they are updated. We believe that this is consistent with the general expectation around `ConfigMap`s. As such, any change to a `ConfigMap` causes it to be replaced (i.e. the current `ConfigMap` is deleted and a new one created).
- The Kubernetes provider requires that any resources that have a user-specified name be deleted before their replacement is created. Normally, Pulumi would schedule the creation of the new resource before the deletion of the original resource in order to avoid downtime. That is not possible for resources with user-specified names, as the new resource would conflict with the original resource.
- The Pulumi engine's delete semantics require that all resources that transitively depend on a resource A are deleted before resource A is deleted. This behavior is required in order to accommodate resources that cannot be deleted until their dependent resources are deleted (there are a number of such resources in AWS), and because the engine does not have granular knowledge about which specific resources might fail in such a fashion. Normally this not an issue: the vast majority of resource deletions do not need to be scheduled promptly, and can be handled at the end of a `pulumi update`. If, however a resource A is being deleted as part of a delete-before-replace, the deletions of A's transitive dependents are scheduled at the same time as the deletion of A. This causes downtime and deletions that a user is unlikely to expect.

We believe that it may be reasonable to avoid the cascading delete by default and instead allow resources to opt in to cascading deletes if necessary, and are investigating whether or not we can make this change without negatively affecting existing scenarios.

**Comment by joeduffy** [comment]
As an aside, I know at least one user was surprised by our decision on ConfigMaps. Since this is just mounted and some apps poll for changes, replacement isn't always desirable. I realize this decision is slightly orthogonal to what's happening here, but are we sure it's the right decision for real world apps?

**Comment by hausdorff** [comment]
I am open to re-considering, but I am pretty confident that this a better for nearly all real-world applications:

* **It's one of the most requested features of Kubernetes.** See [this issue](https://github.com/kubernetes/kubernetes/issues/22368).
* **It allows users to take advantage of the built-in lifecycle semantics of `Pod`.** Triggering a rollout ties the changes in `ConfigMap` (which usually configures, _e.g._, nginx) to the well-understood liveness and readniness semantics of the `Pod`. Using the live-update feature means `Pod`s will simply update whenever the kubelet syncs, and you may never be alerted that something went wrong.
* **It's safer when you decide you need to roll back.** There are no roll-back semantics with the live-update feature, so if you change something, and if you notice something is going wrong, and you have no way to keep directing traffic to old instances.
* **It's safer when the roll-forward fails, and you want to re-direct traffic back to the still-alive replicas from the previous version.** We keep the old `ConfigMap` around, so if roll-forward fails, you not only have access to the old version of the application, but that old version doesn't accidentally get live-updated, because it still references the old `ConfigMap`.
* **It's provides fine-grained controls over _how_ changes roll-out.** Users have access to all the knobs on `Deployment`, or they can roll their own fairly easily with naked `ReplicaSet`s. In contrast, the live-update feature will simply update any `Pod` whenever TTL cache expires and the kubelet syncs.
* **IME it is what the _vast_ majority of users expect.** I have never met someone who encountered this and was _not_ surprised. See also the hundreds of responses in the issue above.
* **IME people use the live-update feature because it's easier, rather than because it's right.** people using live-update are doing it because it's simply easier than the alternative, which is to hand-roll a pipeline for triggering rollouts from `ConfigMap` changes. In all of the cases I have seen, except one, I believe the "correct" way is to go route Pulumi goes instead.
* **The times when you _do_ want to use live-update are almost always better-suited to different technology.** For example, if you want Envoy to update continuously, it's probably better (more scalable, more operationally predictable, safer) for it to continuously watch a config service rather than have the kubelet use a `ConfigMap`.

As I said, I'm always super open to cases where I've missed something, but as it stands now, these to me are very compelling reasons to keep things as they are.

**Comment by geekflyer** [comment]
Just wanted to add to the discussion that a couple of weeks ago there was a bug in pulumi which didn't trigger this rollout behaviour for secrets (fixed by @hausdorff meanwhile) and as a workaround for this I computed a hash of my secret content and added it as annotation to the dependent deployment, which would trigger a new rollout of a deployment. Here's some code which demonstrates this: https://github.com/pulumi-packs/pulumi-packs-typescript/blob/master/incubator/gitlab-runner-on-gke/src/gitlab-kubernetes-runner.ts#L154 .
Afaik this way of using annotations is similar to what helm does.
That being said I have no particular opinion on how pulumi should do it, but I just wanted to add my workaround / "solution" above for completeness to the discussion.

As @hausdorff correctly said the current replacement semantics of pulumi indeed have the advantage that they allow for a cleaner rollout / rollback of deployments and their corresponding configmaps.

**Comment by hausdorff** [comment]
We initially considered going the hashing route, but my thought was that it makes more sense to customize the replacement semantics using annotations on the `ConfigMap` instead of the `Deployment`. So, _e.g._, you might add something like `{"pulumi.com/replacementSemantics": "in-place"}` or something to the `ConfigMap`.

I tend to prefer this approach because I believe it's easier to reason about when the semantics of updating a `ConfigMap` are captured on the `ConfigMap` object itself.

That reminds me of another point I forgot to make, and in the context of Pulumi, it is perhaps the most important:

* **Defaulting to immutable `ConfigMap` makes update semantics explicit in `pulumi preview`.** Replacing the `ConfigMap` on change ensures that we see _which_ deployments are affected by an update.

This last point is actually probably the strongest argument for the current approach. It is extremely doubtful to me that there will be a counter-point that is as impactful as this point.

**Comment by joeduffy** [comment]
I appreciate the detailed reply. As an aside, not entirely related to this thread, I'd also encourage you to write it down somewhere more permanent both for FAQ reasons (I know at least one person has asked) in addition to design archaeology. Perhaps as a Markdown file in pulumi-kubernetes repo?

Given the desire to delete-before-replace, I'm personally at a loss as to what the proposed resolution to this issue might be. As @pgavlin points out, and I'm guessing @swgillespie would agree, the engine seems to be behaving by-design here. It's not legal for resources to have dangling resource dependencies, and if something must be deleted before it is recreated, then by definition don't we need to delete and recreate the transitive graph of resources? And isn't that exactly what you want to happen here for all of the cascading rollout scenarios you're discussing? Or are you asking for a new feature that would allow a provider to hook into the lifecycle here and do something more intelligent, such as setting a property to null, then recreating the resource, and then mutating the resource?

**Comment by hausdorff** [comment]
> It's not legal for resources to have dangling resource dependencies, and if something must be deleted before it is recreated, then by definition don't we need to delete and recreate the transitive graph of resources?

The proposal on the table from @pgavlin is to change the model to distinguish between situations where it is and is not sensible to cascade DBR through all descendants, and to make this distinction on a per-resource basis. This is similar to how the provider distinguishes between fields that require replacement and those that do not.

Not sure about other clouds, but k8s is basically designed to allow users to _entirely decouple_ the lifecycles of separate microservices and resources, and I believe cascading DBR in this case is almost always not what the user expects. There's the example above, but there are others, too. In stateful apps it is very common to DBR the service that depends on it if you make changes to the core state layer. You basically drain traffic, kill the `Pod`s, make the change to the `PersistentVolume`, and then restart the `Pod`s which reattach to the storage layer. (This is basically how the `StatefulSet` rollout works in many cases.) The proposition was that we re-deploy, from scratch, every service that depends on this single stateful layer (including, say, redis caches that will largely prevent requests from reaching the PV layer) is really a steep proposition, for benefits that I personally don't see.

**Comment by hausdorff** [comment]
BTW it seems that this is a sensitive issue... I should say that I'm not trying to step on toes, I'm just trying to explain what I think a good story on k8s means, why, and what the plan is, as I understand it. @pgavlin @lukehoban let me know if I've left stuff out.

**Comment by joeduffy** [comment]
Got it, I think I understand now. The proposal is to make alternative, domain- and resource-specific behavior opt-in, which the Kubernetes provider would then leverage to not do the transitive tear-down. That makes sense, and I think we agreed @pgavlin is on the hook for a design proposal here.

**Comment by hausdorff** [comment]
@joeduffy This seems like a good opportunity to broaden the discussion to the community. I've added to the content calendar.

**Comment by pgavlin** [comment]
Here's a summary of the current situation and the proposed solution:

# Delete-before-replace

## Delete Semantics
When deleting a resource R, Pulumi requires that all resources that depend on R
be deleted before R. This is motivated by two situations:

- Certain resources prevent certain dependencies from being deleted. For example,
  an AWS VPC cannot be deleted until all EC2 instances, subnets, security groups,
  etc. that are associated with that VPC are deleted.
- Certain resources will fail to delete if particular dependencies are deleted
  before the resource itself. For example, an AWS security group rule will fail to
  delete if the security group with which it is associated is deleted before it.

Pulumi opts to capture these relationships as edges in its resource dependency
graph.

When processing deletes, Pulumi uses the dependency graph to ensure that all
dependencies of a given resource are deleted before that resource. Assuming that
all relationships that would prevent deletions to proceed are captured as
dependency edges, this approach is trivially correct. It is, however, only
conservatively correct, as the dependency graph may contain edges that do not
capture one of the delete dependencies described above. In practice, the vast
majority of dependency edges do not capture the delete dependencies described
above: instead, they capture data or runtime dependencies between resources.
Indeed, while there are significant examples of these delete dependencies, they are
rare overall.

In the most common cases--delete-after-replace and stack destruction--the
conservative approach is not an issue in that the approximate set of resources that
must be deleted is close to the actual set of resources that must be deleted. In
the latter case, all resources are trivially dead and this property is obvious. In
the former case, Pulumi transitions all dependents of the replaced resource to the
new physical resource. Dependents may then decide whether or not they also need to
be replaced due to the changes to their inputs. As long as that decision correctly
considers the delete dependencies described earlier, this will result in a correct
and accurate approxmation of the set of resources that must be deleted.

There is one additional case to consider: some resources replacement semantics
require that the current physical resource is deleted before the new physical
resource is created. This is referred to as delete-before-replace.

## Existing Delete-Before-Replace Semantics
The obvious difference in delete-before-replace handling with respect to
delete-after-replace is the timing of the delete. This timing requirement has an
important and non-obvious consequence when the usual delete semantics are applied:
in order process the delete at the correct time, the delete semantics require that
all transitive dependents of the resource to be deleted are deleted beforehand.
Aside from being surprising, this can have disastrous consequences when the set of
transitive dependencies of a resource is not a close approximation of the set of
resources that would prevent the physical resource from being deleted.
Unfortunately, this is the common case, as most edges in the Pulumi dependency
graph are due to dataflow. This over-estimation can cause large swaths of a stack
to be unnecessarily deleted.

## Proposed Delete-Before-Replace Semantics
The "most correct" solution for this issue would be to retain the existing
delete-before-replace semantics, but allow edges to be labeled as delete
dependencies and consider only these edges when scheduling deletes. This is not
necessarily practical, as this may be a difficult decision to make locally: in the
limit it depends on the types of resources an edge connects and the particular
properties involved in the dependency. We believe that it is more practical to
instead adjust the delete-before-replace semantics such that the usual delete
scheduling is not applied. Instead, a resource may opt-in to the delete scheduling
as necessary. We believe that most resources will not opt in to this scheduling, as
most resources do not have delete dependencies. Overall, we believe that this will
deliver the best experience for the largest number of programs, as the impact of a
delete-before-replace resource will be minimized in the most common cases.

In summary, we propose the following:
- Do not schedule deletions of dependents during a delete-before-replace by default
- Allow resources to opt-in to deletion of dependents on a per-type and a per-instance
  basis (i.e. in the resource provider and in the SDK)

**Comment by joeduffy** [comment]
I'm concerned about the proposed approach. You IMHO rightfully pointed out the "most correct" solution and, when it comes to infrastructure management, it would seem we want to be conservative and err on the side of correctness. Furthermore, the idea that most resources can tolerate dependencies being removed out from underneath them seems counter-intuitive to me.

Moreover, CloudFormation, Terraform, and all other such infrastructure as code solutions seem to have chosen our current behavior, as far as I know. I also prefer that we learn from others rather than invent. In fact, isn't this Terraform's precisely behavior for all resource replacements by default?

It's also not clear to me that it's any more or less practical to chase down the tail of resources that cannot tolerate dependencies being deleted without their knowledge, as it is chasing down the tail of resources that can. Indeed, given prior art, I'd assume the former is actually more work.

Surely I am missing a key piece of data/information in why we ended up with this proposal?

**Comment by lukehoban** [comment]
@joeduffy Are you commenting on what the default should be, or on the idea of allowing this to be configured at all?  That is - if the proposal was this instead - would you have the same feedback:

* Allow resources to **opt-out** of deletion of dependents on a per-type and a per-instance
basis (i.e. in the resource provider and in the SDK)

I'm not actually proposing that (yet) - but want to make sure I understand your feedback.

**Comment by joeduffy** [comment]
That's right. I assumed we would land on configurability, but erring on the side of correctness, and letting resources opt out when we know it is safe to do so.

**Comment by joeduffy** [comment]
I also don't know whether you'd want to let a resource make a decision about its dependents. In general, this would seem to be something that only those dependents would be qualified to decide.

**Comment by lukehoban** [comment]
> I also don't know whether you'd want to let a resource make a decision about its dependents. In general, this would seem to be something that only those dependents would be qualified to decide.

So the argument for this is twofold:
1. In the cloud provider, there has to be code in the implementation of resource A that knows to prevent it from being deleted if there are certain kinds of dependencies on it.  It's a property local to the implementation of resource A that we are trying to model.  If I take a dependency on a resource A in Azure from a resource B in AWS, there is no way that can prevent A from being deleted (and thus require this cascading delete-before-replace).
2. In our existing model, we don't have very good ways of understanding what kinds of dependencies have been taken between resources (i.e. via which properties), so it's hard to know how dependents would be able to usefully make this decision about whether they require the cascaded delete.  It would also be much harder (I'm not even sure how) to figure out what needed to be annotated with this information - since the fanout can be huge.

The one big advantage though would be that the cascading delete could be more constrained to only those dependencies that require it - instead of to everything that happened to reference it.

I'd be interested in seeing a concrete proposal for how we could do this via dependents, but I suspect it's very hard to to in our current system - and the proposal above, though more coarse grained, is already much more flexible than the current system.

**Comment by pgavlin** [comment]
> In general, this would seem to be something that only those dependents would be qualified to decide.

That's right. I tried to make this point in the doc with the discussion of "delete edges".

My original proposal was that we allow edge labeling, but that's a rather drastic change. I suspect that there are many cases--certainly much of Kubernetes, as per @hausdorff--where a resource knows that its dependents cannot prevent the resource from being deleted.

**Comment by pgavlin** [comment]
> Moreover, CloudFormation, Terraform, and all other such infrastructure as code solutions seem to have chosen our current behavior, as far as I know. I also prefer that we learn from others rather than invent. In fact, isn't this Terraform's precisely behavior for all resource replacements by default?

This may not in fact be TF's behavior. Consider the following config:
```hcl
resource "aws_s3_bucket" "b" {
  bucket = "my-tf-test-bucket-abcde"
  acl    = "private"

  tags {
    Name        = "My bucket"
    Environment = "Dev"
  }
}

resource "aws_s3_bucket" "c" {
  bucket = "my-tf-test-bucket-efghi"
  acl    = "private"

  tags {
    Name        = "My bucket"
    Environment = "${aws_s3_bucket.b.id}"
  }
}
```

If I deploy this, then change the name of the first bucket, then run an `apply`, the second bucket is not replaced:
```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place
-/+ destroy and then create replacement

Terraform will perform the following actions:

-/+ aws_s3_bucket.b (new resource required)
      id:                          "my-tf-test-bucket-abcde" => <computed> (forces new resource)
      acceleration_status:         "" => <computed>
      acl:                         "private" => "private"
      arn:                         "arn:aws:s3:::my-tf-test-bucket-abcde" => <computed>
      bucket:                      "my-tf-test-bucket-abcde" => "my-tf-test-bucket-abcdef" (forces new resource)
      bucket_domain_name:          "my-tf-test-bucket-abcde.s3.amazonaws.com" => <computed>
      bucket_regional_domain_name: "my-tf-test-bucket-abcde.s3.amazonaws.com" => <computed>
      force_destroy:               "false" => "false"
      hosted_zone_id:              "Z3AQBSTGFYJSTF" => <computed>
      region:                      "us-east-1" => <computed>
      request_payer:               "BucketOwner" => <computed>
      tags.%:                      "2" => "2"
      tags.Environment:            "Dev" => "Dev"
      tags.Name:                   "My bucket" => "My bucket"
      versioning.#:                "1" => <computed>
      website_domain:              "" => <computed>
      website_endpoint:            "" => <computed>

  ~ aws_s3_bucket.c
      tags.%:                      "" => <computed>
      tags.Environment:            "my-tf-test-bucket-abcde" => ""
      tags.Name:                   "My bucket" => ""


Plan: 1 to add, 1 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_s3_bucket.b: Destroying... (ID: my-tf-test-bucket-abcde)
aws_s3_bucket.b: Destruction complete after 1s
aws_s3_bucket.b: Creating...
  acceleration_status:         "" => "<computed>"
  acl:                         "" => "private"
  arn:                         "" => "<computed>"
  bucket:                      "" => "my-tf-test-bucket-abcdef"
  bucket_domain_name:          "" => "<computed>"
  bucket_regional_domain_name: "" => "<computed>"
  force_destroy:               "" => "false"
  hosted_zone_id:              "" => "<computed>"
  region:                      "" => "<computed>"
  request_payer:               "" => "<computed>"
  tags.%:                      "" => "2"
  tags.Environment:            "" => "Dev"
  tags.Name:                   "" => "My bucket"
  versioning.#:                "" => "<computed>"
  website_domain:              "" => "<computed>"
  website_endpoint:            "" => "<computed>"
aws_s3_bucket.b: Creation complete after 7s (ID: my-tf-test-bucket-abcdef)
aws_s3_bucket.c: Modifying... (ID: my-tf-test-bucket-efghi)
  tags.Environment: "my-tf-test-bucket-abcde" => "my-tf-test-bucket-abcdef"
aws_s3_bucket.c: Modifications complete after 6s (ID: my-tf-test-bucket-efghi)

Apply complete! Resources: 1 added, 1 changed, 1 destroyed.
```

**Comment by lukehoban** [comment]
@pgavlin What does Terraform do in the cases where a cascading replace would be needed?  Do these just fail?

**Comment by joeduffy** [comment]
Also curious to hear what @jen20 thinks about this.

**Comment by pgavlin** [comment]
> @pgavlin What does Terraform do in the cases where a cascading replace would be needed? Do these just fail?

Looking now.

**Comment by lblackstone** [comment]
```
  ~ aws_s3_bucket.c
      tags.%:                      "" => <computed>
      tags.Environment:            "my-tf-test-bucket-abcde" => ""
      tags.Name:                   "My bucket" => ""
```
Not sure if it's relevant, but it looks like their plan logic may be a little buggy. The plan suggests that it will set the `tags.Environment` and `tags.Name` to `""`, but the apply actually uses the computed value as I'd expect, and doesn't touch `tags.Name` at all:
```
aws_s3_bucket.c: Modifying... (ID: my-tf-test-bucket-efghi)
  tags.Environment: "my-tf-test-bucket-abcde" => "my-tf-test-bucket-abcdef"
```

**Comment by pgavlin** [comment]
So TF appears to use force-new properties to decide this. For example, consider this config:
```hcl
resource "aws_vpc" "main" {
  cidr_block = "10.0.0.0/16"
}

resource "aws_security_group" "allow_all" {
  name        = "allow_all"
  description = "Allow all inbound traffic"
  vpc_id      = "${aws_vpc.main.id}"

  ingress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }

  egress {
    from_port       = 0
    to_port         = 0
    protocol        = "-1"
    cidr_blocks     = ["0.0.0.0/0"]
  }
}
```
If I apply this config, taint `aws_vpc.main`, and apply the config again, TF will decide to replace the SG and the VPC b/c the `vpc_id` property of the SG forces a new resource. It will then schedule those deletions based on the dependency graph:
```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

-/+ aws_security_group.allow_all (new resource required)
      id:                                   "sg-09ed52b491a19b3cd" => <computed> (forces new resource)
      arn:                                  "arn:aws:ec2:us-east-1:153052954103:security-group/sg-09ed52b491a19b3cd" => <computed>
      description:                          "Allow all inbound traffic" => "Allow all inbound traffic"
      egress.#:                             "1" => "1"
      egress.482069346.cidr_blocks.#:       "1" => "1"
      egress.482069346.cidr_blocks.0:       "0.0.0.0/0" => "0.0.0.0/0"
      egress.482069346.description:         "" => ""
      egress.482069346.from_port:           "0" => "0"
      egress.482069346.ipv6_cidr_blocks.#:  "0" => "0"
      egress.482069346.prefix_list_ids.#:   "0" => "0"
      egress.482069346.protocol:            "-1" => "-1"
      egress.482069346.security_groups.#:   "0" => "0"
      egress.482069346.self:                "false" => "false"
      egress.482069346.to_port:             "0" => "0"
      ingress.#:                            "1" => "1"
      ingress.482069346.cidr_blocks.#:      "1" => "1"
      ingress.482069346.cidr_blocks.0:      "0.0.0.0/0" => "0.0.0.0/0"
      ingress.482069346.description:        "" => ""
      ingress.482069346.from_port:          "0" => "0"
      ingress.482069346.ipv6_cidr_blocks.#: "0" => "0"
      ingress.482069346.prefix_list_ids.#:  "0" => "0"
      ingress.482069346.protocol:           "-1" => "-1"
      ingress.482069346.security_groups.#:  "0" => "0"
      ingress.482069346.self:               "false" => "false"
      ingress.482069346.to_port:            "0" => "0"
      name:                                 "allow_all" => "allow_all"
      owner_id:                             "153052954103" => <computed>
      revoke_rules_on_delete:               "false" => "false"
      vpc_id:                               "vpc-07c766246d1633a68" => "${aws_vpc.main.id}" (forces new resource)

-/+ aws_vpc.main (tainted) (new resource required)
      id:                                   "vpc-07c766246d1633a68" => <computed> (forces new resource)
      arn:                                  "arn:aws:ec2:us-east-1:153052954103:vpc/vpc-07c766246d1633a68" => <computed>
      assign_generated_ipv6_cidr_block:     "false" => "false"
      cidr_block:                           "10.0.0.0/16" => "10.0.0.0/16"
      default_network_acl_id:               "acl-0898383a09d25cb26" => <computed>
      default_route_table_id:               "rtb-02e41af8c9c04b240" => <computed>
      default_security_group_id:            "sg-0159097cabefcefa9" => <computed>
      dhcp_options_id:                      "dopt-ea21cb8c" => <computed>
      enable_classiclink:                   "false" => <computed>
      enable_classiclink_dns_support:       "false" => <computed>
      enable_dns_hostnames:                 "false" => <computed>
      enable_dns_support:                   "true" => "true"
      instance_tenancy:                     "default" => "default"
      ipv6_association_id:                  "" => <computed>
      ipv6_cidr_block:                      "" => <computed>
      main_route_table_id:                  "rtb-02e41af8c9c04b240" => <computed>


Plan: 2 to add, 0 to change, 2 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_security_group.allow_all: Destroying... (ID: sg-09ed52b491a19b3cd)
aws_security_group.allow_all: Destruction complete after 2s
aws_vpc.main: Destroying... (ID: vpc-07c766246d1633a68)
aws_vpc.main: Destruction complete after 1s
aws_vpc.main: Creating...
  arn:                              "" => "<computed>"
  assign_generated_ipv6_cidr_block: "" => "false"
  cidr_block:                       "" => "10.0.0.0/16"
  default_network_acl_id:           "" => "<computed>"
  default_route_table_id:           "" => "<computed>"
  default_security_group_id:        "" => "<computed>"
  dhcp_options_id:                  "" => "<computed>"
  enable_classiclink:               "" => "<computed>"
  enable_classiclink_dns_support:   "" => "<computed>"
  enable_dns_hostnames:             "" => "<computed>"
  enable_dns_support:               "" => "true"
  instance_tenancy:                 "" => "default"
  ipv6_association_id:              "" => "<computed>"
  ipv6_cidr_block:                  "" => "<computed>"
  main_route_table_id:              "" => "<computed>"
aws_vpc.main: Creation complete after 6s (ID: vpc-044c4e5704e360b19)
aws_security_group.allow_all: Creating...
  arn:                                  "" => "<computed>"
  description:                          "" => "Allow all inbound traffic"
  egress.#:                             "" => "1"
  egress.482069346.cidr_blocks.#:       "" => "1"
  egress.482069346.cidr_blocks.0:       "" => "0.0.0.0/0"
  egress.482069346.description:         "" => ""
  egress.482069346.from_port:           "" => "0"
  egress.482069346.ipv6_cidr_blocks.#:  "" => "0"
  egress.482069346.prefix_list_ids.#:   "" => "0"
  egress.482069346.protocol:            "" => "-1"
  egress.482069346.security_groups.#:   "" => "0"
  egress.482069346.self:                "" => "false"
  egress.482069346.to_port:             "" => "0"
  ingress.#:                            "" => "1"
  ingress.482069346.cidr_blocks.#:      "" => "1"
  ingress.482069346.cidr_blocks.0:      "" => "0.0.0.0/0"
  ingress.482069346.description:        "" => ""
  ingress.482069346.from_port:          "" => "0"
  ingress.482069346.ipv6_cidr_blocks.#: "" => "0"
  ingress.482069346.prefix_list_ids.#:  "" => "0"
  ingress.482069346.protocol:           "" => "-1"
  ingress.482069346.security_groups.#:  "" => "0"
  ingress.482069346.self:               "" => "false"
  ingress.482069346.to_port:            "" => "0"
  name:                                 "" => "allow_all"
  owner_id:                             "" => "<computed>"
  revoke_rules_on_delete:               "" => "false"
  vpc_id:                               "" => "vpc-044c4e5704e360b19"
aws_security_group.allow_all: Creation complete after 5s (ID: sg-037f07c7a93036669)

Apply complete! Resources: 2 added, 0 changed, 2 destroyed.
```

The key piece of information that TF has but we are missing is the ability to determine which properties of a dependent resource will change when its dependency is recreated _before_ its dependency is deleted. We are unable to do this, as we do not encode dependencies at a property-level granularity in the checkpoint file.

Given this, I believe that the best path forward is as follows:
- Label each dependency of a resource with the input properties the dependency affects. This involves changing the language provider interface s.t. it can report this information.
- Mark any of a resource's input properties that will force replacement if they change due to the delete-before-replacement of a dependency that affects that property. This involves changing the resource provider interface s.t. it can report this information.

This information will allow the engine to calculate which resources must be replaced due to the delete-before-replacement of a dependency using only information that is statically available. We can only depend on statically available information because we do not have the entire new dependency graph until the Pulumi program has finished executing. This does mean that we may recreate more resources than are strictly necessary (i.e. we may replace a dependent that the program no longer reports).

Importantly, we can still make a conservatively correct decision if the information above is not present in the checkpoint file by assuming that every dependency affects every property and that all properties require replacement.

This is a largish change, but I believe that the adjustment to the model and the required work are quite clear. I do believe that we might be able to get this in during M19.

**Comment by swgillespie** [comment]
This all sounds reasonable to me so far. The language and provider interface changes sounds like they can be done in a backwards-compatible manner. I typed up my intuition about how to do this before I realized that you've already done this:

> Importantly, we can still make a conservatively correct decision if the information above is not present in the checkpoint file by assuming that every dependency affects every property and that all properties require replacement.

Which I 100% agree with. So, sounds good to me.

> This is a largish change, but I believe that the adjustment to the model and the required work are quite clear. I do believe that we might be able to get this in during M19.

Let me know if I can help with bringing this to Python.

**Comment by lukehoban** [comment]
> Mark any of a resource's input properties that will force replacement if they change due to the delete-before-replacement of a dependency that affects that property. This involves changing the resource provider interface s.t. it can report this information.

Could you provide more detail on exactly what you have in mind here.  What does "mark" mean?  Which part of the system computes this, and from what information?

**Comment by lukehoban** [comment]
> Moreover, CloudFormation, Terraform, and all other such infrastructure as code solutions seem to have chosen our current behavior, as far as I know. 

BTW - For CloudFormation, they have create-before-replace as the only option as far as I can tell.  This means they are more like what we do already.  I suspect there must be resources they can't do this for - where they do a delete-before-replace - but at least their docs don't mention it as a possibility and I'm not sure of a specific case to test where we would expect delete-before-replace to be strictly necessary.  I would be interested in finding an example of if they can try to do this, and if so, what their behaviour is.  That said, I would bet if they can end up in this situation, they most likely do similar to Terraform - since they have the same information available: (1) the full graph with edges labelled by properties (2) knowledge of which property edges require replacement.

**Comment by pgavlin** [comment]
> > Mark any of a resource's input properties that will force replacement if they change due to the delete-before-replacement of a dependency that affects that property. This involves changing the resource provider interface s.t. it can report this information.
> 
> Could you provide more detail on exactly what you have in mind here.  What does "mark" mean?  Which part of the system computes this, and from what information?

The thought was that the resource provider would return this as part of the result of `Create`. This is not strictly necessary, however: we can just call `Diff` when computing the set of deletes that need to occur for a delete-before-replace root. This is probably a better choice, as it does not require changes to the provider interface (and therefore works nicely with older providers).

**Comment by naveensrinivasan** [comment]
@lukehoban When is this likely to be fixed? This is a blocker for us. cc @casey-robertson

**Comment by joeduffy** [comment]
@naveensrinivasan We apologize for the delay here. This is a subtle and tricky issue, so it's taken a bit of time to discuss and design. That said, it's in our current sprint, and we will be acting on this soon. Expect a more detailed update tomorrow on the plans and timeframe on a fix.

**Comment by naveensrinivasan** [comment]
Great! We are being blocked by the EKS issue referred here.

**Comment by casey-robertson** [comment]
Checking in on this issue?

**Comment by lukehoban** [comment]
The PR for this is in progress at https://github.com/pulumi/pulumi/pull/2369.  Currently expecting this to be available next week (hopefully early next week).

**Comment by lukehoban** [comment]
This is fixed with #2369 being merged.  We'll still need to release a new CLI and update downstream to pick up the improvements here.
